packages/perseus/src/widget-ai-utils/interactive-graph/prompt-utils.ts

```diff
@@ -0,0 +1,295 @@
+};
+
+type AngleGraphOptions = BaseGraphOptions & {
+    angleOffsetDegrees: number;
+    startCoords?: [Coord, Coord, Coord];
+};
+
+type CircleGraphOptions = BaseGraphOptions & {
+    startParams: {
+        center?: Coord;
+        radius?: number;
+    };
+};
+
+type LinearGraphOptions = BaseGraphOptions & {
+    startCoords?: CollinearTuple;
+};
+
+type LinearSystemGraphOptions = BaseGraphOptions & {
+    startCoords?: CollinearTuple[];
+};
+
+type PointGraphOptions = BaseGraphOptions & {
+    numPoints?: number | "unlimited";
+    startCoords?: Coord[];
+};
+
+type PolygonGraphOptions = BaseGraphOptions & {
+    match?: string;
+    numSides?: number | "unlimited";
+    startCoords?: Coord[];
+};
+
+type QuadraticGraphOptions = BaseGraphOptions & {
+    startCoords?: [Coord, Coord, Coord];
+};
+
+type RayGraphOptions = BaseGraphOptions & {
+    startCoords?: CollinearTuple;
+};
+
+type SegmentGraphOptions = BaseGraphOptions & {
+    numSegments?: number;
+    startCoords?: CollinearTuple;
+};
+
+type SinusoidGraphOptions = BaseGraphOptions & {
+    startCoords?: Coord[];
+};
+
+type GraphOptions =
+    | AngleGraphOptions
+    | CircleGraphOptions
+    | LinearGraphOptions
+    | LinearSystemGraphOptions
+    | PointGraphOptions
+    | PolygonGraphOptions
+    | QuadraticGraphOptions
+    | RayGraphOptions
+    | SegmentGraphOptions
+    | SinusoidGraphOptions;
+
+type AngleUserInput = {
+    coords: [Coord, Coord, Coord];
+    angleOffsetDegrees?: number;
+};
+
+type CircleUserInput = {
+    center: Coord;
+    radius: number;
+};
+
+type LinearUserInput = {
+    coords: CollinearTuple;
+};
+
+type LinearSystemInput = {
+    coords?: CollinearTuple[] | null;
+};
+
+type PointUserInput = {
+    coords?: Coord[] | null;
+};
+
+type PolygonUserInput = {
+    coords: Coord[] | null;
+};
+
+type QuadraticUserInput = {
+    coords: [Coord, Coord, Coord] | null;
+};
+
+type RayUserInput = {
+    coords: CollinearTuple | null;
+};
+
+type SegmentUserInput = {
+    coords: CollinearTuple[] | null;
+};
+
+type SinusoidUserInput = {
+    coords: Coord[] | null;
+};
+
+type UserInput =
+    | AngleUserInput
+    | CircleUserInput
+    | LinearUserInput
+    | LinearSystemInput
+    | PointUserInput
+    | PolygonUserInput
+    | QuadraticUserInput
+    | RayUserInput
+    | SegmentUserInput
+    | SinusoidUserInput;
+
+export type InteractiveGraphPromptJSON = {
+    type: "interactive-graph";
+    options: {
+        graph: {
+            type: string;
+        };
+        backgroundImageUrl: string | null | undefined;
+        range: [min: number, max: number][];
+        labels: ReadonlyArray<string>;
+    };
+    userInput: UserInput;
+};
+
+export const getPromptJSON = (
+    props: React.ComponentProps<typeof interactiveGraph.widget>,
+    userInput: PerseusGraphType,
+): InteractiveGraphPromptJSON | UnsupportedWidgetPromptJSON => {
+    let input: UserInput;
+    let graphOptions: GraphOptions;
+
+    switch (props.graph.type) {
```

Comment:
**thought**: If you extract these two `switch` statements into their own helper functions, could you have them return the value directly from each `case`?  That would avoid you needing all the `as XyzType`, I think. It would also make this function slightly more succinct. YMMV. :)